### PR TITLE
feat: [CDS-45020]: support headers for capability NG http step

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
This change is to support headers in capability check for NG http tasks. This has been rolled out behind a FF. The default behavior is to use headers. 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42886)

### CODE REVIEW

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42886)* status: Some of the checks are too old. 6 file(s) to review. 0 of 2 approvals obtained. 

___
*[`932-connector-task/src/main/java/io/harness/connector/task/tas/TasValidationHandler.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/42886#-NMSC38j3OdmJSLm5LWH:-NMSC38j3OdmJSLm5LWI:bdtuhds) ([raw file](https://github.com/harness/harness-core/blob/4f45da6aa860ce4d6459f9fb32d9114d09c3c3d1/932-connector-task/src/main/java/io/harness/connector/task/tas/TasValidationHandler.java#L68)):*
> ```Java
>   private ConnectorValidationResult handleValidateTask(CloudFoundryConfig cfConfig, String endpointUrl) {
>     try {
>       boolean valid = Http.connectableHttpUrlWithoutFollowingRedirect(endpointUrl, new ArrayList<>());
> ```

logic to pass empty headers inside the service



_Code quote (from 930-delegate-tasks/src/main/java/io/harness/delegate/task/executioncapability/HttpConnectionExecutionCapabilityCheck.java):_
```Java
isIgnoreRedirect()
```
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/42886#-NMSD6jy1Xh2VMAWD8VD:-NMSD6jy1Xh2VMAWD8VE:b-1b7ryi) ([raw file](https://github.com/harness/harness-core/blob/4f45da6aa860ce4d6459f9fb32d9114d09c3c3d1/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java#L44)):*
> ```Java
>     List<KeyValuePair> headers = new ArrayList<>();
>     if (EmptyPredicate.isNotEmpty(requestHeader)) {
>       for (HttpHeaderConfig headerConfig : requestHeader) {
> ```

logic to include headers behind FF
___
*[`980-commons/src/main/java/io/harness/network/Http.java` line 149 at r1](https://reviewable.io/reviews/harness/harness-core/42886#-NMSD-Ja8gJqG23PGiv8:-NMSD-Ja8gJqG23PGiv9:bi6zy41) ([raw file](https://github.com/harness/harness-core/blob/4f45da6aa860ce4d6459f9fb32d9114d09c3c3d1/980-commons/src/main/java/io/harness/network/Http.java#L149)):*
> ```Java
>                 connection.setInstanceFollowRedirects(false);
> 
>                 // set headers
> ```

logic behind FF.

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java [hunk 1]
```diff
@@ -33,10 +36,22 @@ public class HttpTaskParametersNg implements TaskParameters, ExecutionCapability
   int socketTimeoutMillis;
   boolean useProxy;
   boolean isCertValidationRequired;
+  boolean shouldAvoidHeadersInCapability;
 
   @Override
   public List<ExecutionCapability> fetchRequiredExecutionCapabilities(ExpressionEvaluator maskingEvaluator) {
+    if (shouldAvoidHeadersInCapability) {
```
Comment (yogesh-chauhan) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/http/HttpTaskParametersNg.java:43]: lets add test for this

File: 980-commons/src/main/java/io/harness/network/Http.java [hunk 2]
```diff
@@ -121,13 +121,13 @@ public Integer load(String url) throws IOException {
             }
           });
 
-  LoadingCache<String, Integer> responseCodeForValidationWithoutFollowingRedirect =
+  LoadingCache<HttpURLHeaderInfo, Integer> responseCodeForValidationWithoutFollowingRedirect =
```
Comment (yogesh-chauhan) [inline 980-commons/src/main/java/io/harness/network/Http.java:124]: lets add test for this

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 1 of 6 files at r1, 2 of 3 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42886)* status: 10 file(s) to review. 1 of 2 approvals obtained.

**Comment by hingerabhinav** [comment]
trigger ft1, ft2

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 2 of 3 files at r2, 5 of 7 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42886)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42886)* status: :shipit: complete! ✓

**Pull request merged**